draft-gpew-priv-ppm.md

```diff
@@ -947,26 +947,50 @@ aggregator has validated and which fall in the batch interval of the request.
-
-Aggregate requests are ordered as follows: We say that a report `R2` follows
-report `R1` if either `R2.time > R1.time` or `R2.time == R1.time` and `R2.nonce
-> R1.nonce`. If `R2.time < R1.time`, or `R2.time == R1.time` but `R2.nonce <=
-R1.nonce`, then we say that `R2` does not follow `R1`.
-
-To prevent replay attacks, each aggregator ensures that each report it
-aggregates follows the previously aggregated report. To prevent the adversary
-from tampering with the ordering of reports, honest clients incorporate the
-ordering-sensitive parameters `(time, nonce)` into the AAD for HPKE encryption.
-Note that this strategy may result in dropping reports that happen to have the
-same timestamp and nonce value.
+about the client's measurement, violating the privacy goal of PPM. To
+prevent such replay attacks, this specification requires the aggregators to
+detect and filter out replayed reports.
+
+To detect replay attacks, each aggregator keeps track of the set of nonces
+pertaining to reports that were previously processed for a given task. If the
+leader receives a report from a client whose nonce is in this set, it simply
+ignores it. Similarly, a helper ignores an encrypted input share whose nonce is
+in this set.
+
+A malicious aggregator may attempt to force a replay by replacing the nonce
+generated by the client with a nonce its peer has not yet seen. To prevent this,
+clients incorporate the nonce into the AAD for HPKE encryption, ensuring that
+the input share is only recovered if the aggregator is given the correct nonce.
+(See {{upload-request}}.)
 
 Aggregators prevent the same report from being used in multiple batches (except
 as required by the protocol) by only responding to valid collect requests, as
 described in {{batch-parameter-validation}}.
 
+The size of the nonce set is unbounded, and depending the task, can grow quite
+large. To ensure they do not need to store these indefinitely, each aggregator
+MAY at its own discretion ignore reports with timestamps that are too far in the
+past, too far in the future, or would fall in a batch interval that has already
+been aggregated (see {{batch-parameter-validation}}).
+
+[OPEN ISSUE: Allowing aggregators to ignore reports at-will is likely to lead
+to reports getting dropped unnecessarily. Ideally there would be a way to force
+the aggregators to make the same decision about the time intervals for which
+nonces are tracked and for which nonces are ignored. There are two ways to this:
+
+1. Pick a duration `d_min` for reports too far in the future and a duration
+   `d_max` for reports too far in the past, and ignore a report if `now() - time
+   > d_dmin` or `time - now() > d_max`, where `time` is the report timestamp.
+   This requires the aggregators to have a roughly synchronized clock.
+
+2. @martinthomson made the following suggestion, which doesn't require clock
+   sync. The leader and helper could periodically negotiate the time `t_min` in
```

Comment:
Why is it useful not to require clock sync? The leader and the helper are servers, and they're quite capable of having accurate clocks.

Reply:
Good sync != perfect sync.  Aggregation will fail if anti-replay is independently enforced by helpers (which is something I argue we want) and any amount of relative skew could produce disagreements.

Reply:
> The leader and the helper are servers, and they're quite capable of having accurate clocks.

I tend to agree. If we align `d_max` and `d_min` with `min_batch_duration`, then all they have to do is ensure their clocks are within `min_batch_duration` seconds of one another. However I think @martinthomson's broader point is this requirement isn't strictly necessary. The true requirement is that they agree on what set of nonces they have to track at any given time.

This was discussed extensively here: https://mailarchive.ietf.org/arch/msg/ppm/oDCOJ-cUR8NQFG6At3byH1fdcd4/

Reply:
@martinthomson AFAICT a disagreement only results needlessly dropping reports, I don't see how it can result in a replay (i.e., privacy violation). I'm probably missing something though.

Reply:
I'm not sure what you mean by "perfect sync". NTP is capable of providing 10ms error. Given that these timestamps are in seconds, this seems fine. I don't see a real problem here.

Reply:
I agree with @martinthomson here -- it's best if we can simply avoid skew altogether, and that seems doable with the agreed upon min/max limits.

Reply:
This seems much more complicated than simply having sync.

Reply:
I'm not convinced is as easy as that. Would you be OK with merging without a change and moving the discussion to an issue? (As I proposed above.)

Reply:
I don't really see how it would be complicated. This is the same as aggregators agreeing on batch interval boundaries, no?